Native label scan store as default

tinwelint opened the pull request:
Contains some changes which hindered a full green build with NativeLabelScanStore as the default label scan store:

- One noticeable change is that import-tool no longer builds the label scan store as part of the import, which affects users by delaying the next startup of the db (where the label scan store automatically gets built), but also making import faster. Read more about why this was done in that specific commit.
- Another noticeable change is the setting for users to override which label scan store to use for the db, `dbms.label_index`. Setting is required and defaults to `NATIVE` Most users will never need to touch this setting though. 
- The presence of NativeLabelScanStore made the InMemoryLabelScanStore (used in testing) redundant and so it was also removed.

This PR depends on https://github.com/neo4j/neo4j/pull/8698 since that PR fixes an issue where a checkpoint in the middle of a copy of the native label scan store may corrupt it. It would be wise to merge that PR first.

cl: [kernel] Storage of labels are generally faster and uses less memory due to a new and native label index. This index is set as the default, but can be changed using `dbms.label_index` setting. No migration is required, but the new label index will be automatically built on the next start-up of a db.

tinwelint committed:
Doesn't build label scan store as part of import

this because:

 - environment/config isn't necessarily the same for import, which means there
   may be a different page size or even a different LSS configured for the import
 - performance of node import is impacted if building LSS concurrently since
   I/O will be random and shared between node store and LSS

The label scan store will be built on first db startup, where the correct config
is used and so the correct lss will be built.

tinwelint committed:
Properly uses correct LSS in whole CC main/test

also faster FullCheckIntegrationTest due to not starting and stopping full db
for every "given" transaction to apply.

tinwelint committed:
Corrected store file count expectation in a test

tinwelint committed:
Native LSS is allowed to create its store even in read-only node

tinwelint committed:
LuceneLabelScanStore gets told if it's read-only

previously it was always writable, no matter what the neo4j config said.
Integerestingly compliance testing tested the LuceneLabelScanStore directly
and it asserted that a read-only label scan store shouldn't be able to
create an initial store.

When looking at the rest of the stores, they are actually created, even when
read-only mode is set. Label scan store just happened to not fail when it was
in read-only mode. So now lss aligns with stores in general in that they can
be created even in read-only mode.

tinwelint committed:
Fixes an issue in primitive base iterator

where additional calls to hasNext() after it had returned false would continue
to delegated to fetchNext().

tinwelint committed:
Sets native label scan store as default

At the same time changes the label scan store setting

- Renames it to dbms.label_index Because to the user it's an index,
  an internal setting though since normally there's no need to change it.
- Sets a default value "native" for that setting. This is because we want
  to pin down which implementation is used instead of selecting highest prioritized
  every time, which may result in different implementations - depending on environement.

Also removed InMemoryLabelScanStore. This was mostly there because the main implementation
of the label scan store lived in a module "above" kernel and so kernel tests needed its
own for testing a full database. Now the native label scan store lives inside kernel and so
the in-memory test version is redundant.

MishaDemianenko commented on community/neo4j/src/test/java/upgrade/StoreUpgraderTest.java:
```diff
@@ -154,6 +153,13 @@ public StoreUpgraderTest( String version )
     }
 
     @Before
```
now we have two `@Before` methods :)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -480,13 +481,15 @@
-            "label scan store is set by the kernel extension loading it." )
-    @Internal
-    public static final Setting<String> label_scan_store = setting( "unsupported.dbms.label_scan_store", STRING, NO_DEFAULT );
+    public static enum LabelIndex
```
static modifier is redundant

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:
```diff
@@ -480,13 +481,15 @@
+        LUCENE;
+    }
+
+    @Description( "Backend to use for label --> nodes index" )
```
do we really want this setting to be public? why users should have possibility to choose implementation? maybe we can make it internal/unsupported?

tinwelint replied:
Typically users won't need to care since it will "just work", but at this point in time it's slightly optimistic. Perhaps we could still have it as `@Internal` (which makes it not show up publically, right?)

MishaDemianenko replied:
yes i think so, vote for `@Internal`

MishaDemianenko commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanIndexBuilder.java:
```diff
@@ -71,4 +71,9 @@ public LabelScanIndex build()
         return isReadOnly() ? new ReadOnlyDatabaseLabelScanIndex( format, storageBuilder.build() )
                             : new WritableDatabaseLabelScanIndex( format, storageBuilder.build() );
```
this line probably should invoke` buildWritable()` now instead

MishaDemianenko commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:
```diff
@@ -114,9 +116,22 @@ else if ( !luceneIndex.isValid() )
     @Override
     public void start() throws IOException
     {
+        System.out.println( "LUCENE!!" );
```
debug output

MishaDemianenko commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:
```diff
@@ -36,16 +35,19 @@
 {
-    private final LabelScanIndex luceneIndex;
+    private final LuceneLabelScanIndexBuilder indexBuilder;
+    private volatile LabelScanIndex luceneIndex;
```
maybe we should rename this field now :)

tinwelint replied:
Uhm, sure... to what and why?

MishaDemianenko replied:
it's not a lucene index anymore right? maybe just to `index` or `labeIndex`?

tinwelint replied:
It's a field in the `LuceneLabelScanStore` so I hope it has to do with lucene :)

MishaDemianenko replied:
aaaaaaa, true :)

MishaDemianenko commented on community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/labelscan/LuceneLabelScanStore.java:
```diff
@@ -114,9 +116,22 @@ else if ( !luceneIndex.isValid() )
         }
     }
 
+    private void create() throws IOException
```
The whole idea before was to have index that are fully responsible for creation, flushing, closing etc to avoid index volatile variable and necessity of multiple index builders and state handling. With new way it's gone and we need to handle possible multiple indexes and always use volatile variable to read something from index. Maybe we can avoid doing that and come back to old way of doing things?

tinwelint replied:
I'm not sure how to do that though

MishaDemianenko replied:
We currently already have `ReadOnlyAbstractDatabaseIndex/WritableAbstractDatabaseIndex` but those expect AbstractLuceneIndex maybe those can be updated to expect something else - but those are really lucene oriented currently so that can be some additional work; 
or instead we can just delegate creation to index always - in that case index will replace it's internal structures, not sure how that will work with current new index

tinwelint replied:
It'll just be a volatile reference somewhere else, won't it?

Reading a volatile which no one ever writes to, except ones at startup, aint so bad (knowing details about the java memory model), is it?

MishaDemianenko replied:
yeah thats true, in previous indexes we have a partitions couple of levels below that where guarded anyway;
ok then it lets keep it for now like it is know; that's true - do not think we will have any problems with it

MishaDemianenko reviewed (changes requested):
Couple of comments, otherwise looks great!

tinwelint committed:
StoreCopyClient injects label index config from parent

tinwelint committed:
LabelScanStoreHaIT can replace label index extension

And does so by only removing all currently loaded label index extensions
before adding the specific extension it wants.

tinwelint committed:
StoreMigration tool adds dependencies to load label index

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/NamedLabelScanStoreSelectionStrategy.java:
```diff
@@ -31,20 +35,26 @@
+ * <li>Specifically configured label scan store name, see {@link GraphDatabaseSettings#label_index},
  * each {@link LabelScanStoreProvider} is instantiated with a name, which is matched with the configured name.
  * The setting can also be left unspecified, at which point the next parameter kicks in:</li>
  * <li>Highest prioritized instance, actually using {@link HighestSelectionStrategy}</li>
```
This javadoc does not look correct anymore. As I understand it, an exception will be thrown if the setting is left unspecified and priority will never be taken into account. Correct?

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/scan/NativeLabelScanStoreExtension.java:
```diff
@@ -55,7 +56,7 @@
     public NativeLabelScanStoreExtension()
     {
-        this( 0 /*disabled by default*/, LabelScanStore.Monitor.EMPTY );
+        this( 100, LabelScanStore.Monitor.EMPTY );
```
If I've understood it correctly priority is never taken into account when picking `LabelScanStore` and so priority could be removed from both Native and Lucene `...LabelScanStoreExtension`.

tinwelint replied:
True, we can probably remove this altogether. I'll try that!

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreChaosIT.java:
```diff
@@ -50,6 +50,6 @@ private static File storeFile( File directory )
     protected void addSpecificConfig( GraphDatabaseBuilder builder )
     {
-        builder.setConfig( label_scan_store, NativeLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
+        builder.setConfig( label_index, NativeLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
```
Would be nice to use the `LabelIndex` enum from `GraphDatabaseSettings` explicitly.

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreStartupIT.java:
```diff
@@ -28,14 +28,14 @@
     protected void addSpecificConfig( GraphDatabaseBuilder builder )
     {
-        builder.setConfig( label_scan_store, NativeLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
+        builder.setConfig( label_index, NativeLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
```
`LabelIndex` enum

burqen commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanStoreUpdateIT.java:
```diff
@@ -38,7 +38,7 @@
         protected void configure( GraphDatabaseBuilder builder )
         {
-            builder.setConfig( label_scan_store, NativeLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
+            builder.setConfig( label_index, NativeLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
```
`LabelIndex` enum

burqen commented on community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreChaosIT.java:
```diff
@@ -82,6 +82,6 @@
     protected void addSpecificConfig( GraphDatabaseBuilder builder )
     {
-        builder.setConfig( label_scan_store, LuceneLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
+        builder.setConfig( label_index, LuceneLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
```
`LabelIndex` enum

burqen commented on community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreStartupIT.java:
```diff
@@ -33,15 +33,15 @@
     protected void addSpecificConfig( GraphDatabaseBuilder builder )
     {
-        builder.setConfig( label_scan_store, LuceneLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
+        builder.setConfig( label_index, LuceneLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
```
`LabelIndex` enum

burqen commented on community/lucene-index/src/test/java/org/neo4j/graphdb/LuceneLabelScanStoreUpdateIT.java:
```diff
@@ -36,7 +36,7 @@
         protected void configure( GraphDatabaseBuilder builder )
         {
-            builder.setConfig( label_scan_store, LuceneLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
+            builder.setConfig( label_index, LuceneLabelScanStoreExtension.LABEL_SCAN_STORE_NAME );
```
`LabelIndex` enum

burqen commented on community/lucene-index/src/test/java/org/neo4j/index/recovery/UniqueIndexRecoveryTests.java:
```diff
@@ -92,7 +94,15 @@ public void before()
+        return (GraphDatabaseAPI) factory
+                .newEmbeddedDatabaseBuilder( storeDir.absolutePath() )
+                .setConfig( GraphDatabaseSettings.label_index, LabelIndex.LUCENE.name() )
+                .newGraphDatabase();
```
Should this test also work with `Native`? It looks like it. Maybe make it abstract and have one implementation for `Lucene` and one for `Native`, like with the `LabelScanStore` specific tests.

tinwelint replied:
That's really a test of recovery of indexes and uniqueness constraints for the lucene schema index provider. Adding a native variant wouldn't really add anything. It's however unclear why the lucene label index is used specifically, but doesn't seem to matter.

burqen reviewed (changes requested):
Looks good in general, got some small comments.

tinwelint committed:
Removes priority from LabelScanStoreProvider

Since this selection is done explicitly from config instead

tinwelint committed:
Refers to LabelIndex setting directly

burqen commented:
@tinwelint Approved! I'll wait for feedback from @MishaDemianenko and CI before merging.
